sys/wtimer/wtimer_core.c

```diff
@@ -0,0 +1,466 @@
+            select_list_last = long_list_head;
+            long_list_head = long_list_head->next;
+        }
+    }
+
+    /* cut the "selected long timer list" at the end */
+    if (select_list_last) {
+        select_list_last->next = NULL;
+    }
+
+    /* merge "current timer list" and "selected long timer list" */
+    if (timer_list_head) {
+        if (select_list_last) {
+            /* both lists are non-empty. merge. */
+            timer_list_head = _merge_lists(timer_list_head, select_list_start);
+        }
+        else {
+            /* "selected long timer list" is empty, nothing to do */
+        }
+    }
+    else { /* current timer list is empty */
+        if (select_list_last) {
+            /* there's no current timer list, but a non-empty "selected long
+             * timer list".  So just use that list as the new current timer
+             * list.*/
+            timer_list_head = select_list_start;
+        }
+    }
+}
+
+/**
+ * @brief handle low-level timer overflow, advance to next short timer period
+ */
+static void _next_period(void)
+{
+#if WTIMER_MASK
+    /* advance <32bit mask register */
+    _high_cnt += ~WTIMER_MASK + 1;
+    if (! _high_cnt) {
+        /* high_cnt overflowed, so advance >32bit counter */
+        _long_cnt++;
+    }
+#else
+    /* advance >32bit counter */
+    _long_cnt++;
+#endif
+
+    /* swap overflow list to current timer list */
+    timer_list_head = overflow_list_head;
+    overflow_list_head = NULL;
+
+    _select_long_timers();
+
+}
+
+/**
+ * @brief main wtimer callback function
+ */
+static void _timer_callback(void)
+{
+    uint32_t next;
+    uint32_t reference;
+
+    DEBUG("_timer_callback() now=%u (%u)pleft=%u\n", (unsigned)wtimer_now(),
+            (unsigned)_mask(wtimer_now()), (unsigned) _mask(0xffffffff-wtimer_now()));
+
+    if (!timer_list_head) {
+        DEBUG("_timer_callback(): tick\n");
+        /* there's no timer for this timer period,
+         * so this was a timer overflow callback.
+         *
+         * In this case, we advance to the next timer period.
+         */
+        _next_period();
+
+        reference = 0;
+
+        /* make sure the timer counter also arrived
+         * in the next timer period */
+        while (_wtimer_now() == _mask(0xFFFFFFFF));
+    }
+    else {
+        /* we ended up in _timer_callback and there is
+         * a timer waiting.
+         */
+        /* set our period reference to that timer's target time. */
+        reference = _mask(timer_list_head->target);
+    }
+
+overflow:
+    /* check if next timers are close to expiring */
+    while (timer_list_head && _time_left(_mask(timer_list_head->target), reference) < WTIMER_ISR_BACKOFF) {
+        /* make sure we don't fire too early */
+        while (_time_left(_mask(timer_list_head->target), 0));
+
+        _shoot(timer_list_head);
+
+        /* advance to next timer in list */
+        timer_list_head = timer_list_head->next;
```

Comment:
if the timer callback fired in the previous `shoot()` call immediately re-sets the timer using say `wtimer_set()`, then the `timer_list_head->next` will most likely have been reused and be invalid. In fact the timer referenced by `timer_list_head` should probably not be considered valid anymore after the `_shoot()` call.